### PR TITLE
Prevent accidental file matching when glob contains trailing slash

### PR DESCRIPTION
**Pull request by luketomlinson**
Resolves #268 

Handles an odd case with `Minimatch` that we worked around before that inadvertently would match files when a trailing slash was added.

### CODE REVIEW

File: packages/glob/src/internal-pattern.ts [hunk 1]
```diff
@@ -132,7 +137,7 @@ export class Pattern {
       // Append a trailing slash. Otherwise Minimatch will not match the directory immediately
       // preceding the globstar. For example, given the pattern `/foo/**`, Minimatch returns
       // false for `/foo` but returns true for `/foo/`. Append a trailing slash to handle that quirk.
-      if (!itemPath.endsWith(path.sep)) {
+      if (!itemPath.endsWith(path.sep) && !this.implicitTrailing) {
```
Comment (ericsciple) [inline packages/glob/src/internal-pattern.ts:140]: implicit globstar

File: packages/glob/src/internal-globber.ts [hunk 2]
```diff
@@ -102,6 +110,8 @@ export class DefaultGlobber implements Globber {
       stack.unshift(new SearchState(searchPath, 1))
     }
 
+
```
[the remark this answers is not in the document]
Comment (ericsciple) [reply packages/glob/src/internal-globber.ts:113]: 😆

File: packages/glob/src/internal-pattern.ts [hunk 3]
```diff
@@ -47,13 +47,16 @@ export class Pattern {
   // Disable no-dupe-class-members due to false positive for method overload
   // https://github.com/typescript-eslint/typescript-eslint/issues/291
 
+  private readonly implicitTrailing: boolean
+
   constructor(pattern: string)
   constructor(pattern: string, segments: undefined, homedir: string)
-  constructor(negate: boolean, segments: string[])
+  constructor(negate: boolean, segments: string[], homedir?: string, implicitTrailing?: boolean)
```
Comment (ericsciple) [inline packages/glob/src/internal-pattern.ts:54]: isImplicitPattern

**Review by ericsciple — APPROVED** [review]

File: packages/glob/src/internal-pattern.ts [hunk 4]
```diff
@@ -144,7 +162,10 @@ export class Pattern {
 
     // Match
     if (this.minimatch.match(itemPath)) {
-      return this.trailingSeparator ? MatchKind.Directory : MatchKind.All
+      const matchValue = this.trailingSeparator
```
Comment (TingluoHuang) [inline packages/glob/src/internal-pattern.ts:165]: why do we need to create a const?

Comment (luketomlinson) [reply packages/glob/src/internal-pattern.ts:165]: good call. debugging left in

**Review by thboop — APPROVED** [review]
LGTM

**Pull request merged**